raft.go

```diff
@@ -1988,64 +1949,85 @@ func (r *Raft) shouldSnapshot() bool {
 	return delta >= r.conf.SnapshotThreshold
 }
 
-// takeSnapshot is used to take a new snapshot.
+// takeSnapshot is used to take a new snapshot. This must only be called from
+// the snapshot thread, never the main thread.
 func (r *Raft) takeSnapshot() error {
 	defer metrics.MeasureSince([]string{"raft", "snapshot", "takeSnapshot"}, time.Now())
-	// Create a snapshot request
-	req := &reqSnapshotFuture{}
-	req.init()
 
-	// Wait for dispatch or shutdown
+	// Create a request for the FSM to perform a snapshot.
+	snapReq := &reqSnapshotFuture{}
+	snapReq.init()
+
+	// Wait for dispatch or shutdown.
 	select {
-	case r.fsmSnapshotCh <- req:
+	case r.fsmSnapshotCh <- snapReq:
 	case <-r.shutdownCh:
 		return ErrRaftShutdown
 	}
 
 	// Wait until we get a response
-	if err := req.Error(); err != nil {
+	if err := snapReq.Error(); err != nil {
 		if err != ErrNothingNewToSnapshot {
 			err = fmt.Errorf("failed to start snapshot: %v", err)
 		}
 		return err
 	}
-	defer req.snapshot.Release()
+	defer snapReq.snapshot.Release()
 
-	// Log that we are starting the snapshot
-	r.logger.Printf("[INFO] raft: Starting snapshot up to %d", req.index)
+	// Make a request for the configurations and extract the committed info.
+	// We have to use the future here to safely get this information since
+	// it is owned by the main thread.
+	configReq := r.getConfigurations()
+	if err := configReq.Error(); err != nil {
+		return err
+	}
+	committed := configReq.configurations.committed
+	committedIndex := configReq.configurations.committedIndex
 
-	// Create a new snapshot
+	// We don't support snapshots while there's a config change outstanding
+	// since the snapshot doesn't have a means to represent this state. This
+	// is a little weird because we need the FSM to apply an index that's
+	// past the configuration change, even though the FSM itself doesn't see
+	// the configuration changes. It should be ok in practice with normal
+	// application traffic flowing through the FSM. If there's none of that
+	// then it's not crucial that we snapshot, since there's not much going
+	// on Raft-wise.
```

Comment:
It seems like processLog() could just write to fsmCommitCh for any type of log entry, and the FSM already filters out uninteresting ones. We'd just have to remove the switch statement out of processLog().